google/cloud/storage/async_writer.h

```diff
@@ -0,0 +1,145 @@
+ * to recover from most network problems between the application and Google
+ * Cloud Storage. The application can buffer any streaming data in memory, and
+ * periodically force a flush in the resumable upload. The service will return
+ * the number of bytes persisted and the application and then discard the
+ * corresponding bytes from its buffer.
+ *
+ * Some caveats apply:
+ * - The application should be prepared for "partial flush" results, that is
+ *   service may indicate than fewer bytes were persisted vs. the number of
+ *   bytes sent.
+ * - The application can continue to upload data after a flush. It is possible
+ *   that a flush request returns more persisted bytes than were available when
+ *   the flush request is sent.
+ */
+class AsyncWriter {
+ public:
+  AsyncWriter() = default;
+  explicit AsyncWriter(std::unique_ptr<AsyncWriterConnection> impl);
+  ~AsyncWriter();
+
+  ///@{
+  /// @name Move-only.
+  AsyncWriter(AsyncWriter&&) noexcept = default;
+  AsyncWriter& operator=(AsyncWriter&&) noexcept = default;
+  AsyncWriter(AsyncWriter const&) = delete;
+  AsyncWriter& operator=(AsyncWriter const&) = delete;
+  ///@}
+
+  /**
+   * The upload id.
+   *
+   * Applications that need to resume uploads after a restart should checkpoint
+   * this value to persistent storage.
+   */
+  std::string UploadId() const;
+
+  /**
+   * The state of the current upload.
+   *
+   * This returns the last "known" state of the upload. The values are updated
+   * (1) when the `AsyncWriter` object is created, and (2) on calls to `Query()`
+   * and `Finalize()`.
+   *
+   * Applications may finalize an upload, and then try to resume the upload. In
+   * this case this function returns `storage::ObjectMetadata`.
+   *
+   * More commonly, applications resume an upload that is in progress, or
+   * periodically flush the
+   *
+   * During an upload the service will periodically persist the uploaded data.
+   * Applications should not assume that all data "sent" is persisted. The
+   * service may buffer the data in ephemeral storage before persisting it.
+   *
+   * If the application needs to resume an upload that was interrupted, then it
+   * must check `persisted_state()` before sending data. The service will assume
+   * that new data starts at the correct offset.
+   *
+   * If an upload is resumed after it is finalized the library will return a
+   * variant holding `storage::ObjectMetadata` value.
+   *
+   * Otherwise the variant returns the size of the persisted data. The
+   * application should send the remaining data to upload, starting from this
+   * point.
+   */
+  absl::variant<std::int64_t, storage::ObjectMetadata> PersistedState() const;
+
+  /**
+   * Returns the offset for the next `Write()` or `Finalize()` call.
+   *
+   * This is automatically updated once the `Write()` call is satisfied.
+   * Applications may use this information to seek the right data from the
```

Comment:
I'm not sure what "to seek the right data" means.

Reply:
Removed for now